core/src/smartcontracts/isi/permissions/mod.rs

```diff
@@ -0,0 +1,305 @@
+        HasTokenBoxed, IsAllowedBoxed,
+    };
+}
+
+#[cfg(test)]
+mod tests {
+    #![allow(clippy::restriction)]
+
+    use std::{collections::BTreeSet, str::FromStr as _};
+
+    use iroha_data_model::{expression::prelude::*, isi::*};
+
+    use super::{builder::Validator as ValidatorBuilder, combinators::DenyAll, *};
+    use crate::wsv::World;
+
+    #[derive(Debug, Clone, Serialize)]
+    struct DenyBurn;
+
+    impl From<DenyBurn> for IsInstructionAllowedBoxed {
+        fn from(permissions: DenyBurn) -> Self {
+            Box::new(permissions)
+        }
+    }
+
+    impl IsAllowed<Instruction> for DenyBurn {
+        fn check(
+            &self,
+            _authority: &AccountId,
+            instruction: &Instruction,
+            _wsv: &WorldStateView,
+        ) -> Result<()> {
+            match instruction {
+                Instruction::Burn(_) => Err("Denying sequence isi.".to_owned().into()),
+                _ => Ok(()),
+            }
+        }
+    }
+
+    #[derive(Debug, Clone, Serialize)]
+    struct DenyAlice;
+
+    impl IsAllowed<Instruction> for DenyAlice {
+        fn check(
+            &self,
+            authority: &AccountId,
+            _instruction: &Instruction,
+            _wsv: &WorldStateView,
+        ) -> Result<()> {
+            if authority.name.as_ref() == "alice" {
+                Err("Alice account is denied.".to_owned().into())
+            } else {
+                Ok(())
+            }
+        }
+    }
+
+    impl From<DenyAlice> for IsInstructionAllowedBoxed {
+        fn from(value: DenyAlice) -> Self {
+            Box::new(value)
+        }
+    }
+
+    #[derive(Debug, Clone, Serialize)]
+    struct GrantedToken;
+
+    // TODO: ADD some Revoke tests.
+
+    impl HasToken for GrantedToken {
+        fn token(
+            &self,
+            _authority: &AccountId,
+            _instruction: &Instruction,
+            _wsv: &WorldStateView,
+        ) -> std::result::Result<PermissionToken, String> {
+            Ok(PermissionToken::new(
+                Name::from_str("token").expect("Valid"),
+            ))
+        }
+    }
+
+    fn asset_id(
+        asset_name: &str,
+        asset_domain: &str,
+        account_name: &str,
+        account_domain: &str,
+    ) -> IdBox {
+        IdBox::AssetId(AssetId::new(
+            AssetDefinitionId::new(
+                asset_name.parse().expect("Valid"),
+                asset_domain.parse().expect("Valid"),
+            ),
+            AccountId::new(
+                account_name.parse().expect("Valid"),
+                account_domain.parse().expect("Valid"),
+            ),
+        ))
+    }
+
+    #[test]
+    pub fn multiple_validators_combined() {
+        let permissions_validator: IsInstructionAllowedBoxed =
+            ValidatorBuilder::with_validator(DenyBurn)
+                .with_validator(DenyAlice)
+                .all_should_succeed()
+                .build();
```

Comment:
isn't  `all_should_succeed` a terminating operation? Do we need explicit call to `build` following it?

Reply:
I think it was you who suggested to introduce `build()`?